### ISSUE TITLE
crate_universe broken with bzlmod

### ISSUE THREAD

**Issue by jondo2010** [bazelbuild/rules_rust#2483]
This regressed in #2461 

At  https://github.com/AmeliasCode/rules_rust/blob/13ba928eddf944ff6f8636f53f5f869e8b01894d/crate_universe/extension.bzl#L49 breaks most uses of `rules_rust` in bzlmod since it hard codes the repository name to `~override`. This works in CI since the `rules_rust` module in the tests are all overridden.

**Comment by UebelAndre** [comment]
cc @AmeliasCode @illicitonion

**Comment by AmeliasCode** [comment]
@UebelAndre @jondo2010 Any ideas on how best to get the canonical repository name for the generated repositories? If I understand correctly, the actual repositories are created here: https://github.com/AmeliasCode/rules_rust/blob/13ba928eddf944ff6f8636f53f5f869e8b01894d/crate_universe/extension.bzl#L131 but the BUILD files are generated before-hand.

**Comment by AmeliasCode** [comment]
I'm thinking something like this (names will be improved later): 

```
    lbl = Label(":target")
    idx = str(lbl).find("//")
    rep = str(lbl)[:idx]
    rendering_config = json.decode(render_config(
        regen_command = "Run 'cargo update [--workspace]'",
        crate_label_template = "{}~crate~{{repository}}__{{name}}-{{version}}//:{{target}}".format(rep),
    ))
```

but this seems kind of hacky? It works, but I'm still not sure where `~crate~` is coming from, or if this would have other unintended consequences.

**Comment by AmeliasCode** [comment]
>I'm still not sure where ~crate~ is coming from

Answering my own question here, it's the name of the extension.

**Comment by cameron-martin** [comment]
Why does the canonical name need to be used here, rather than utilising the repo mappings?

**Comment by AmeliasCode** [comment]
@cameron-martin Can you expand on that a bit? I'm fairly new to Bazel

**Comment by cameron-martin** [comment]
Bzlmod uses a feature called "repo mappings" to map non-canonical repository names to canonical repository names. I'm not sure that the form of the canonical repository names is designed to be a stable public API, so it would be better to specify non-canonical repository names and let bazel do this mapping.

The issue here is that since labels are passed as strings to the repository using `all_crate_deps`, this resolution of non-canonical to canonical repository names is done from the perspective of this repository, which doesn't have the necessary repo mappings. I think the correct fix here is to render the labels in `@crates//:defs.bzl` wrapped in a `Label` constructor so that this resolution is done from the perspective of the `@crates` repository, which does have the necessary repo mappings.

**Comment by cameron-martin** [comment]
I have put up an alternative fix here: #2491

**Comment by AmeliasCode** [comment]
That's the approach I was looking for! Thanks, Cameron!
